src/window.c

```diff
@@ -742,6 +765,82 @@ l_window_warpMouse(lua_State *L)
 	return 0;
 }
 
+#if SDL_VERSION_ATLEAST(2, 0, 4)
+
+typedef struct {
+	lua_State *L;	// The Lua state the callback is in
+	int ref;	// The registry index of the function to call
+} CallbackData;
+
+static SDL_HitTestResult
+hitTestCallback(SDL_Window *win, const SDL_Point *area, CallbackData *cd)
+{
+	SDL_HitTestResult ht;
+	int st = lua_gettop(cd->L);
+
+	lua_geti(cd->L, LUA_REGISTRYINDEX, cd->ref);
+	commonPush(cd->L, "p", WindowName, win);
+	videoPushPoint(cd->L, area);
+
+	/* If the callback errored or didn't return a value, we get HITTEST_NORMAL */
+	ht = luaL_optinteger(cd->L, -1, SDL_HITTEST_NORMAL);
+
+	lua_settop(cd->L, st);
+
+	return ht;
+}
+
+/*
+ * Window:setHitTest(func)
+ *
+ * Arguments:
+ *	func a callback function or nil to clear the hit-test callback.
+ */
+static int
+l_window_setHitTest(lua_State *L)
+{
+	SDL_Window *w 	= commonGetAs(L, 1, WindowName, SDL_Window *);
+	int t		= lua_type(L, 2);
+	int ut;
+
+	/* Clear a window's existing callback */
+	if ((ut = lua_getuservalue(L, 1)) == LUA_TUSERDATA) {
+		CallbackData *cd = *(CallbackData **)lua_touserdata(L, -1);
+		luaL_unref(cd->L, LUA_REGISTRYINDEX, cd->ref);
+		lua_pop(L, 1);
+		lua_pushnil(L);
+		lua_setuservalue(L, 1);
+	}
+	else {
+		lua_pop(L, 1);
+	}
+
+	if (t == LUA_TFUNCTION) {
+		CallbackData *cd;
+		if ((cd = malloc(sizeof (CallbackData))) == NULL)
```

Comment:
Because SDL can call the callback at any time, and I don't want the callback data going near garbage collection.  Personal preference.

Also, I'm a bit fuzzy on whether it is okay to dereference a Lua userdata pointer after the userdata is off the stack, which it has the potential to be.